### ISSUE TITLE
Supporting Apple Silicon

### ISSUE THREAD

**Issue by kevzhao2** [MonoMod/MonoMod#90]
I spent some time debugging over the weekend and figured out the necessary changes for supporting Apple Silicon. I can try to restructure the code a bit and send in a pull request, but I just wanted to write down the things that I encountered.

## W^X
Apple requires all pages to be W^X, and requires pages to be mapped in with the `MAP_JIT` flag.
1. When applying detours, the relevant pages are already mapped in with the `MAP_JIT` flag. We just need to do the following:
    - Call `pthread_jit_write_protect_np(false)` (to enable write and disable execute)
    - Make the changes
    - Call `pthread_jit_write_protect_np(true)` (to disable write and enable execute)
1. When creating the trampoline for the JIT hook, we need to `mmap` in memory with the `MAP_JIT` flag instead of calling the equivalent of `malloc`.

The first item has to be done entirely in native code because calling `pthread_jit_write_protect_np(false)` from managed code causes a bus error -- this is because `pthread_jit_write_protect_np` controls _every_ page!

## libclrjit.dylib
`libclrjit.dylib` doesn't seem to be one of the modules loaded into the process, which means that the JIT hooks don't work. This is, however, a pretty easy change to make.

## Virtual Method Function Pointers
Hooking virtual methods seems to be a bit fraught (with arm64, at least?) because the function pointers that we get with `DetourRuntimeNETPlatform.GetFunctionPointer` don't seem to related to the entries found in the vtable, and I'm not exactly sure why this isn't the case for amd64...

My quick and dirty solution was to just read the vtable to get the function pointer:
```csharp
var methodTable = method.DeclaringType.TypeHandle.Value;
var methodDesc = method.MethodHandle.Value;

var m_wSlotNumber = (ushort) Marshal.ReadInt16((IntPtr) (methodDesc.ToInt64() + 4));
var m_wFlags = (ushort) Marshal.ReadInt16((IntPtr) (methodDesc.ToInt64() + 6));
if ((m_wFlags & 0x8000) == 0) {
    m_wSlotNumber &= 0xff;
}

var chunkNumber = m_wSlotNumber / 8;
var chunkOffset = m_wSlotNumber % 8;

var vtablePtr = Marshal.ReadIntPtr((IntPtr) (methodTable.ToInt64() + 0x40 + 8 * chunkNumber));
var vtableSlotPtr = (IntPtr) (vtablePtr.ToInt64() + 8 * chunkOffset);

ptr = Marshal.ReadIntPtr(vtableSlotPtr);
```

Additionally, because the devirtualization code expects the entries to point to things that look like precode, I had to change the detour code a bit to match.

**Comment by nike4613** [comment]
> W^X stuff
I find it hard to believe that the only way to adjust protections is to adjust the protections for *every* JIT-mapped page. Does `mprotect`ing the relevant pages with the correct permissions not work?

> `libclrjit.dylib`
Where is the JIT loaded then? As far as I'm aware, CoreCLR always loads the JIT as a separate binary.

> Virtual Functions
`GetFunctionPointer` getting a value different from reading the VTable is expected. Pretty much every managed call in the runtime, until it's fully optimized, goes through *at least* one stub. `GetFunctionPointer` skips through those to find the actual function bodies, and apply the detour there. If virtual calls aren't working, that's likely some other issue which is also present.

**Comment by kevzhao2** [comment]
> > W^X stuff
> > I find it hard to believe that the only way to adjust protections is to adjust the protections for _every_ JIT-mapped page. Does `mprotect`ing the relevant pages with the correct permissions not work?

The `mprotect` calls failed with EACCES. [Apparently this is fairly recent behavior?](https://developer.apple.com/forums/thread/672804)

> > `libclrjit.dylib`
> > Where is the JIT loaded then? As far as I'm aware, CoreCLR always loads the JIT as a separate binary.

I'm.. not really quite familiar enough with what's going on. But this was always `null`, and there only seemed to be a single module loaded:
```csharp
                ProcessModule clrjitModule = currentProc.Modules.Cast<ProcessModule>()
                    .FirstOrDefault(m => Path.GetFileNameWithoutExtension(m.FileName).EndsWith("clrjit", StringComparison.Ordinal));
```

> > Virtual Functions
> > `GetFunctionPointer` getting a value different from reading the VTable is expected. Pretty much every managed call in the runtime, until it's fully optimized, goes through _at least_ one stub. `GetFunctionPointer` skips through those to find the actual function bodies, and apply the detour there. If virtual calls aren't working, that's likely some other issue which is also present.

Yeah that makes sense. So really the problem is that we didn't fill out precode walkers here, and the ngen comment is unrelated:
```csharp
            if (PlatformHelper.Is(Platform.ARM)) {
                // TODO: Debug detouring NGEN'd methods on ARM.

            }
```

Shouldn't this have been noticed earlier though??

**Comment by nike4613** [comment]
> Shouldn't this have been noticed earlier though??

Until fairly recently, there simply wasn't a (non-Mono) runtime which targeted ARM and was widely used. As such, ARM walking was never implemented. Now would likely be the time to do it, though I would not blame you at all for not wanting to given the amount of pain that has gone into the x86 walking.

> there only seemed to be a single module loaded

Could you enumerate the modules loaded in a CoreCLR process from outside? On Windows this would be done with something like Process Hacker, but I'm not sure what the equivalents would be for MacOS. It would likely be the same as on Linux. 

> The `mprotect` calls failed with EACCES.

*Sigh.* Personally, I'd like to avoid pulling in more shellcode, for hopefully obvious reasons. Is there any way to change a mapping away from `MAP_JIT`? Otherwise, it may be possible to coerce the JIT to generate code which we could just copy verbatim into another block to perform the detours.

**Comment by 0x0ade** [comment]
Thank you for compiling all that information. I'll copy paste what I've sent someone else on the MonoMod Discord server when they asked about Apple Silicon support earlier this month:

> macOS in general is just close enough™ to Linux, and Android ARM support ended up being "works on my machine, sorry I can't get it working on your machine" as well
> I'll happily try to review any community-maintained efforts at supporting macOS on Apple ARM, but can't start working on it myself, thus have no real plans
> maybe once I'll be a few months into my new job, I'll maybe be able to afford a M1 testing device, but that's a huge *maybe*.
> ... and by then, I'll probably have less free time available than I've had for the past few years.

Having said that, I'll happily review your (and anyone else's) efforts whenever I can, and highly appreciate that you've put your time into this to begin with.

----

> there only seemed to be a single module loaded

... and after checking the dotnet runtime repo, that seems to be an intended workaround, given BSD limitations / lack of standardization..? Compare the following:
- [ProcessManager.BSD.cs](https://github.com/dotnet/runtime/blob/fd33b9eac6f259ef731e0204bcd2aebd534b8ddc/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.BSD.cs#L37)
- [ProcessManager.Linux.cs](https://github.com/dotnet/runtime/blob/fd33b9eac6f259ef731e0204bcd2aebd534b8ddc/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs#L39)

I am only familiar with Linux, where `procfs` is a given. I've recently had to work around this in another project of mine, and had to resort to calling a command line helper and parsing its output. In this scenario, on Linux that'd be `pmap`, on macOS it sounds like it'd be `vmmap`, but I can't test that and don't know how reliably parseable its output is. And more importantly...

> The first item has to be done entirely in native code because calling `pthread_jit_write_protect_np(false)` from managed code causes a bus error -- this is because `pthread_jit_write_protect_np` controls *every* page!

> Personally, I'd like to avoid pulling in more shellcode, for hopefully obvious reasons.

I can't even imagine how we'd embed shellcode like [we already do for icache flushing](https://github.com/MonoMod/MonoMod.Common/blob/7d799091ba6e740988b82fe233cb3fa00ef32611/RuntimeDetour/Platforms/Native/DetourNativeARMPlatform.cs#L227) (**which was only tested on Android and would also require updating, see [LLVM's `clear_cache.c`](https://android.googlesource.com/toolchain/compiler-rt/+/release_36/lib/builtins/clear_cache.c)**), or detour non-JITed code, given how...

> Is there any way to change a mapping away from `MAP_JIT`?

The only way of achieving this that I was made aware of by Hello71 on Discord is to copy it out, unmap it, remap it as MAP_JIT and copy it back. I'd love to get my hands on Apple Silicon to check if that really is necessary - I hope it isn't.

As much as I hate to say it, at this point I'd be willing to let Azure Pipelines / GitHub Actions compile a helper native library, embed it, unpack it onto the disk and load it at runtime, as long as size and per-platform code can be kept to an absolute minimum, and *only if absolutely necessary*.  
I mention embedding because RuntimeDetour can be used in environments where shipping a native library alongside the managed library is *impossible*, and where the managed library is copied between platforms. For example, a mod loader for a game loads mods from .zips. You can load managed .dlls from .zips, but native libs must exist on the disk, the mod loader might not be able to unpack and resolve native dependencies like that, and modders certainly won't create per-platform .zips anyway.  
At least I'd rather ship *that* than "[let's download CLR function offsets from a selfhosted server that parses PDBs](https://www.codeproject.com/Articles/463508/NET-CLR-Injection-Modify-IL-Code-during-Run-time)".

Getting back to the point of actually iterating through loaded modules, [the docs for `mach_vm_region_recurse`](https://developer.apple.com/documentation/kernel/1402114-mach_vm_region_recurse) (which, according to some random comments on the internet™️, is what vmmap might be using) look very... promising. And to me, it still feels like "dotnet on macOS can't list all loaded modules" should be fixed by the dotnet team, no matter how much we'll work around that. Let's first see if that function - or `vmmap` in general - is in any way helpful though. Or maybe `pmap` instead of `vmmap`. Once again, I wish I could get my hands on Apple Silicon.

> and the ngen comment is unrelated:

For some context: I began adding precode walkers when someone reported problems related to ngen'd assemblies, and it's all gotten much complex over time, to the point where RuntimeDetour now needs to works around Wine behavior and I've given up and added a generic catch-all. I should go back and fix those misnomers at some point though.

> Until fairly recently, there simply wasn't a (non-Mono) runtime which targeted ARM and was widely used.

I remember having tested dotnet 2.x on one of my older rooted phones a few years ago, everything seemed fine enough to meet my personal needs, interest was lost eventually, thus yea. RuntimeDetour is heavily underutilized on ARM, both inside and outside mono, and as mentioned previously (icache), I wouldn't be surprised if other parts required updating as well.

**Comment by hakusaro** [comment]
> Having said that, I'll happily review your (and anyone else's) efforts whenever I can, and highly appreciate that you've put your time into this to begin with.
> [...]
> I'd love to get my hands on Apple Silicon to check if that really is necessary

Hi @0x0ade! Nice to meet you. I was actually made aware of this discussion thanks to input from @SignatureBeef. In a conversation with them, I reached the conclusion that providing money or hardware for Apple Silicon is definitely something we _could_ consider. My biggest problem with doing this is that it might create psychological obligations for you, and if you don't have the time/energy to do it, it isn't worth it anyways. It's an option that I should probably put on the table though -- I don't know how much time/energy/etc outside of just the raw issue of getting access to the hardware you want to spend on it. It _sounds_ like you want AS to test on, but it also sounds like you don't realistically have the time. Consider it food for thought: we would love to get arm working _and_ Apple Silicon working, and if having the hardware will unblock you, that's something we should consider. But if it's just going to sit in a closet or create stress for you, I don't want you to shoulder that burden necessarily.

**Comment by 0x0ade** [comment]
If this had happened half a year ago, time wouldn't have been a concern 😅 My first day at my new job is tomorrow. I'll let you know when I'll know if I'll even have the time and energy to work on this in my time off, as I have yet to see what stress awaits me. Thank you for the offer though!

**Comment by kevzhao2** [comment]
I think I can send in a PR to implement some of the precode walking for arm64 some time this week, which should unblock @hakusaro and @SignatureBeef's issues with general arm64 support.

> As much as I hate to say it, at this point I'd be willing to let Azure Pipelines / GitHub Actions compile a helper native library

Embedding the native library seems like the right choice, because I can't see any ways to not use `pthread_jit_write_protect_np`:
1. We could try to save the page, map the page out, map it in again without the `MAP_JIT` flag, and restore the page.

    a. But leaving the page without the `MAP_JIT` flag is problematic if coreclr wants to change something in that page, since it will use `pthread_jit_write_protect_np` instead of `mprotect` for changing the page protections!

    b. ...and mapping it in again with the `MAP_JIT` flag is problematic since we'd have to initialize the page, or somehow have the old values lying around...
2. ...or we could try to map in a new page without the `MAP_JIT` flag, but this would involve us repointing stuff to the new page, which just runs into the same issues.

Also, by embedding a native library, we'll at least be able to use `sys_icache_invalidate`.

> Getting back to the point of actually iterating through loaded modules, [the docs for `mach_vm_region_recurse`](https://developer.apple.com/documentation/kernel/1402114-mach_vm_region_recurse) (which, according to some random comments on the internet™️, is what vmmap might be using) look very... promising. And to me, it still feels like "dotnet on macOS can't list all loaded modules" should be fixed by the dotnet team, no matter how much we'll work around that. Let's first see if that function - or `vmmap` in general - is in any way helpful though. Or maybe `pmap` instead of `vmmap`. Once again, I wish I could get my hands on Apple Silicon.

`vmmap` definitely provides the right output:
```
__TEXT                      105f60000-106130000    [ 1856K  1856K     0K     0K] r-x/rwx SM=COW          /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.2/libclrjit.dylib
__DATA_CONST                106130000-106144000    [   80K    48K    48K    32K] r--/rwx SM=COW          /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.2/libclrjit.dylib
__LINKEDIT                  106160000-1061cc000    [  432K   432K     0K     0K] r--/rwx SM=COW          /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.2/libclrjit.dylib
__DATA                      106144000-106148000    [   16K    16K    16K     0K] rw-/rwx SM=COW          /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.2/libclrjit.dylib
__DATA                      106148000-106160000    [   96K    16K    16K    32K] rw-/rwx SM=PRV          /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.2/libclrjit.dylib
```
No idea how stable this, so I would prefer `mach_vm_*` functions over trying to parse this output, but searching for "libclrjit.dylib" seems reasonable as a last resort.

> My first day at my new job is tomorrow. 

Good luck! Hope everything goes smoothly :)

**Comment by 0x0ade** [comment]
Initial work on ARM64 precode walking by @kevzhao2 merged with 69a2aa1. I'll keep this issue open for further progress tracking.

I've already mentioned this on the Discord server in more detail, but will recap here. I'm currently unable to use "Mac M1 as a service" services because:
- Scaleway hates my credit card, or my credit card hates Scaleway, as Mastercard's "ID check" popup tells me to contact my credit card client service despite it working everywhere else.
- AWS M1 instances are still in preview, I can only request an invitation and hope for the best.
- MacStadium's prices aren't something I feel comfortable randomly throwing money at, and they don't offer free trials.

My current plan is: Given that I'll need a laptop when I'll relocate closer to my new workplace sooner or later anyway, I'll probably save up until I can afford a M1 MacBook, hitting two birds with one stone. This is to be seen as completely independent from MonoMod - getting it won't mean that I'll immediately work on M1 support for MonoMod, and I want to pay it from my own income, not from community donations which could introduce pressure. By the time I'll get it, I'll hopefully have settled a bit more at my new job and will be able to work on this in my spare time at my own pace. If the community ends up creating solid M1 support for MonoMod before that happens, I'll finally be able to review it by then, and continue maintaining it. And if something happens to my new employment (I'm only one week in, and it's going well so far), I'll see how to move forward from there.

**Comment by hakusaro** [comment]
@0x0ade hey! how's the new job? I joined your patreon a while back, but I'd be more than happy to throw some gas on the fire and send you an M1 MacBook to an address you send me if you need one.

**Comment by 0x0ade** [comment]
Saw this pop up in my email inbox again, sorry for the lack of updates, sadly lacking time right now thus keeping it short: I'm relocating next month. All my previous plans were shot down by stuff going on in the household I'm living in right now that I won't dwell much on, which I can't wait to finally get out of.

But I was already very close to buying a M1 MacBook multiple times now. If everything goes well, I might finally press the buy button this or next month.

At this point I can't promise that I'll have the free time to look into M1 support though.

**Comment by hakusaro** [comment]
No problemo, don't let not having the product stop you though. Once you're settled, hit me up!

**Comment by YogurtTheHorse** [comment]
Hey! Just wanted to know how is your progress on getting Silicon Mac and the issue itself? :)

**Comment by nike4613** [comment]
I have finally started working on x64 Mac support on [`reorganize`](https://github.com/MonoMod/MonoMod/tree/reorganize/), after having finally gotten access to one. I'd like to get that working, then I'll take a look at working on Arm, though I don't currently have access to one.

**Comment by GrimLothar** [comment]
Please ignore this comment, I just want to remember in the future that I'm following this issue as it looks like it's a requirement to get [Harmony](https://github.com/pardeike/Harmony/issues/424) working on Silicon, which is also the requirement to get [FastScriptReload](https://github.com/handzlikchris/FastScriptReload/issues/31) working on Silicon. 🙏

**Comment by karanshah-browserstack** [comment]
Any ETA when can we expect this to be released?

**Comment by nike4613** [comment]
There is currently no ETA. I simply don't have the hardware to be able to work on this.

**Comment by hakusaro** [comment]
@nike4613 you can hit me up in the MonoMod discord (i’m antiparticles) for hardware if you’re seriously only limited by hardware for this issue

**Comment by pardeike** [comment]
I could contribute something too if necessary. Especially if I'm not the only one.

**Comment by Nihlus** [comment]
I've run into an instance where I need Harmony on an ARM platform as well, though I'd be happy with any sort of ARM support, not just M1. I'm happy to offer any sort of support I can.

Edit: reading back... MonoMod should have some level of support on ARM already, right? Am I not understanding that correctly? If so, I should perhaps open up a separate issue for the platform I'm currently having trouble with.

**Comment by nike4613** [comment]
It currently does not have any support for ARM.

**Comment by Nihlus** [comment]
@nike4613 Gotcha. How does https://github.com/MonoMod/MonoMod.Common/blob/master/RuntimeDetour/Platforms/Native/DetourNativeARMPlatform.cs relate to this?

**Comment by nike4613** [comment]
That (and all of MMC) are part of legacy MonoMod. Current development (and broadest support) is on the `reorganize` branch, which is a nearly full rewrite, and that does not have any ARM support.

Legacy's ARM support was all "it works on my machine" as-is, so it got dropped along with the rest of legacy, pending a return when I or someone else has the time/resources to work on it.

**Comment by Nihlus** [comment]
Okay, I see - thanks for the explanation :) Is there some way I can support or aid the ARM implementation? I'm not familiar enough with MonoMod's architecture or the ABI of ARM processors to take a crack at it all on my lonesome, but I'm more than willing to help out any way I can.

**Comment by kirilllysenko** [comment]
I would also like to aid in making ARM implementation. I have a hardware to test on, but don't have any experience with MonoMod

**Comment by tyronx** [comment]
Vintage Story dev here. ARM support would enable us to port our game to Mac. I've doubled my patreon pledge. If additional hardware is needed to accelerate development of this, do let me know.

**Comment by MisutaaAsriel** [comment]
Forgive me for asking about this here, but has there been any progress made in this regard? VRChat just released their Unity 2022 SDK in beta, and their scripting tool, Udon Sharp, uses Harmony Lib as a key component. Unity 2022 itself brings with it native Apple Silicon support, which greatly improves performance and stability over running Intel builds in Rosetta. However, this issue in particular seems to effect Harmony, which in turn effects Udon Sharp, which in turn effects the VRChat SDK.

Issues regarding Apple Silicon support for Harmony [point here](https://github.com/pardeike/Harmony/issues/424#issuecomment-1595898404). With a major platform having a shift towards newer software, which provide Apple Silicon support, and said platform having plans for Apple Silicon support themselves in the extended pipeline (iOS/macOS), I fear this issue might become more pressing as time goes on.

**Comment by hakusaro** [comment]
I'm still happy to sponsor hardware for people who want to work on getting this (now, `reorganize`) working on Apple Silicon (including `arm64`/`arm64e`) architectures.

**Comment by Chicken-Bones** [comment]
> I'm still happy to sponsor hardware for people who want to work on getting this (now, `reorganize`) working on Apple Silicon (including `arm64`/`arm64e`) architectures.

Likewise

**Comment by zed-0xff** [comment]
> I'm still happy to sponsor hardware for people who want to work on getting this (now, `reorganize`) working on Apple Silicon (including `arm64`/`arm64e`) architectures.

+1

**Comment by Guztaver** [comment]
> I would also like to aid in making ARM implementation. I have a hardware to test on, but don't have any experience with MonoMod

Me neither, I am interested in running tMod on the Arm64 CPU, so, I can help testing, since I don't know how MonoMod works too

**Comment by Fadenfire** [comment]
I've experimented with this issue to try to figure out what all needs to be done for arm64 support. Some of what I've discovered has already been mentioned here, but there are also some things that I have not seen mentioned that may be helpful for creating a proper port. I've mostly been trying to get my changes to work on .NET core 6, so I don't know how well my changes work for other versions of .NET or other runtimes. Here's a list of my major observations:

1. As mentioned before in this issue, apple silicon requires that pages are never mapped write and exec at the same. It instead requires you to map the memory with `MAP_JIT` and then use `pthread_jit_write_protect_np` to toggle all `MAP_JIT` pages between read-write and read-exec for the current thread. This means that `PatchData` in `MacOSSystem` needs to be modified to use `pthread_jit_write_protect_np` when copying to executable memory. Unfortunately, the memory copy itself can't be done in managed code, as calling `pthread_jit_write_protect_np(0)` will disable execution for ALL `MAP_JIT` pages, which includes the JIT compiled code of `PatchData`. This means trying to use `pthread_jit_write_protect_np` inside of managed code will always result in BAD_ACCESS. I solved this by created a small C++ helper lib with a simple function like this:
    ```C++
    extern "C" void copy_to_jit(void* from, void* to, size_t length)
    {
	    pthread_jit_write_protect_np(0);
    
	    std::memcpy(to, from, length);
    
	    pthread_jit_write_protect_np(1);
    }
    ```
    Which is then called from managed code.

2. MonoMod hooks into `CILJit::compileMethod` by replacing the corresponding entry in the `CILJit` vtable with a pointer to a managed function. However, on apple silicon .NET core sometimes calls `compileMethod` with `pthread_jit_write_protect_np` turned off, preventing the execution of any JIT compiled managed code. This is a problem as any managed hooks that we insert into the vtable will crash as soon as `compileMethod` is called with JIT write protection turned off. The only solution I could find was to create another function in my C++ helper library and then point the vtable entry for `compileMethod` to my C++ helper function. The original `compileMethod` method pointer is placed in a global variable inside of the C++ helper library. When `compileMethod` is called, it'll call my C++ function, which will then use the global variable to call the original `compileMethod` method along with a call back into managed code to allow MonoMod to manipulate the JIT result. My C++ function can be found [here](https://github.com/Fadenfire/MonoMod/blob/89aa2d88a29a4287f8bfae9f8d9a1630a5ff462f/src/MonoMod.Core/Platforms/Architectures/arm64/macos_helper_arm64.cpp#L68-L88).
3. On arm64, .NET core 6 seems to use a similar JIT system as .NET core 7, placing the compiled JIT code into a separate area instead of `entryAddress`. This means that I had to backport the fix in `Core70Runtime` to `Core60Runtime`. Unfortunately, I ran into a similar problem as above, as the `Core70Runtime` fix adds a managed hook to `ICorJitInfo::allocMem`. Since calling any JIT compiled managed code during `CILJit::compileMethod` crashes, this doesn't work for apple silicon. I was able to fix this by using the same strategy as above, by creating another function in my C++ helper library, overwriting the vtable entry of `ICorJitInfo::allocMem` with my C++ function, and then storing the original `allocMem` method in a C++ global variable. There is probably a better way to fix this, but this is what I got to work.

The code for my attempt at adding arm64 support can be found [here](https://github.com/Fadenfire/MonoMod/tree/macos-m1), if anyone's interested.

**Comment by nike4613** [comment]
Nice work.

> 1. ...

I'm glad the new PatchData method interface is proving useful. :)

> 2. ...

This is slightly concerning. I wonder what causes this? 

We already use (if available) a native exception helper stub between native and managed code everywhere we inject ourselves into the JIT/EE boundary. Could we use those stubs to make sure `pthread_jit_write_protect_np` is in the correct state? (Is it possible to query the state?)

> 3. ...

This sounds to me like we should inject another type in between `Core70` and `Core60` (some `Core60WXORXRuntime` maybe? Also worth double checking with the release branch of dotnet/runtime for it, see what's different) to move much of the .NET 7 support code into.

---

For an actual release, I'd like to avoid having C++-compiled binaries if possible, to avoid needing platform-specific toolsets to build those binaries. Currently, we're able to avoid it by using NASM and LLD-Link; I suspect a similar approach with a portable assembler + linker would probably be the approach here, though the amount of code necessary here is rather unfortunate.

**Comment by Fadenfire** [comment]
> We already use (if available) a native exception helper stub between native and managed code everywhere we inject ourselves into the JIT/EE boundary. Could we use those stubs to make sure pthread_jit_write_protect_np is in the correct state?

I'm pretty sure that is an option. The main issue is that currently `PosixExceptionHelper` uses `CreateSpecialEntryStub` to pass the managed function pointer to the exception helper. The generated stub would have to have been created with `MAP_JIT`, which would cause a crash as soon as it is called from native code. This can probably be fixed by using some sort of global state to pass information to the exception helper instead of a generated stub.

>  (Is it possible to query the state?)

From what I can find there isn't a way. (although Apple's documentation is very sparse on the subject) This is especially problematic, as sometimes `CILJit::compileMethod` is called with write protection is disabled and sometimes it's called with it enabled. I just pretended that it was always disabled and re-enabled it accordingly. This seems very fragile though and could lead to crashes later on, so there's probably a better way.

**Comment by nike4613** [comment]
> The generated stub would have to have been created with `MAP_JIT`

Can we not allocate those in a page marked RW, then remap them as RX? We need few enough of them that wasting most of a page is probably not a huge issue.

**Comment by Fadenfire** [comment]
Yeah, that's potentially an option as well. I wasn't sure if Apple Silicon allows remapping a page to be exec, but if it does then that would be a much better option than using global state.

**Comment by allquixotic** [comment]
I have a used but fully functional M1 MacBook Air sitting on my desk, wiped and ready to be activated, and I’m willing to ship it for free to anyone who needs Apple Silicon hardware to do development on this issue. To keep myself from being ripped off by someone lacking the technical chops to actually do the work, I will review your public open source contributions (GitHub or any other public repos of your choice that I can verify you have committed to) before agreeing to send anything. If anyone is serious about taking me up on this offer, please email smcnam@gmail.com or reply here. 

For logistical reasons I can only ship to countries on relatively good terms with NATO.

**Comment by vprodan** [comment]
I did some work over `reorganize` branch to get `ARM64` support. [WIP branch](https://github.com/vprodan/MonoMod/tree/reorganize-arm64)
For some reason I've get `EXC_BAD_ACCESS (code=2, address=0x104bcc060)` on `macOS`, but `linux` passes JIT stage but macOS fails for some reason (applicable only for .net 7, 8. For 6 and above macOS also passes, suppose becouse there is no hooks in MonoMod).
It left to finish `KnownMethodThunks` (I hope this is last step), and also I'm missing `NativeExceptionHelper`
If u know how to fix something of it or maybe have expertise of low-level arm please contact me 😊
UPD. checked @Fadenfire comments, and fork, got some answers. @Fadenfire if u're interested of collaboration of arm64 port, please contact me

**Comment by overdramaticpanromantic** [comment]
I'm not very well-versed in technical stuff, but I wanted to come here to ask - one of the games I want to play relies on MonoMod, and I'm curious as to if there's a rough ETA for compatibility. Apologies for taking up y'all's time.

**Comment by allquixotic** [comment]
> I'm not very well-versed in technical stuff, but I wanted to come here to ask - one of the games I want to play relies on MonoMod, and I'm curious as to if there's a rough ETA for compatibility. Apologies for taking up y'all's time.

There is no ETA. There are not very many developers who have the knowledge and time to work on this specific capability, so the best we can do is wait, unless you are willing to contract someone and pay them for their time.

**Comment by jordanamr** [comment]
Is there any way we can start a bounty on this one? I'm willing to contribute some cash.

**Comment by biltongza** [comment]
I have an M1 Mac, I've been a C# dev for a decade, I think I have a working-ish knowledge of some aspects of native dev though not a lot of experience. I've also never touched this codebase. But, I'm interested in helping out, I figure it would be cool to learn. Plus, I have some incentive, since I'm a Risk of Rain 2 enjoyer which apparently depends on this for mods to work. Where is a good place to start?

**Comment by igor84** [comment]
Regarding avoiding C/C++ code because of platform specific tools required for compiling I understand that people are using Zig as C/C++ cross compilation toolkit. Zig download is 50MB - 80MB depending on the platform and my understanding is that it contains everything you need to cross compile from any platform to any platform (Win/Linux/Mac).

**Comment by JVindenes** [comment]
Ah yeah, so this is where my VRChat on Mac adventure ends... I would also be interested to contribute some cash to be this, but it could be more interesting to contact VRChat, maybe they could help in financing some hours to investigate how much work would need to be done. I'll send them a quick e-mail, it's worth a shot.

**Comment by LeadAssimilator** [comment]
For anyone working on this, like @vprodan it looks like BytePattern will need to some improvements for arm since addresses are bitpacked in bytes along with other operations. Take LDR (literal) for example, Bits 0-4 are Rt while 5-23 are imm19. Unless I'm missing something, BP currently can't both matching against partial bits and extracting the remainder as part of an address. This will be needed for finding the KnownMethodThunks such as FixupPrecode, StubPrecode and CallCountingStub code.

I don't currently have an Mx mac, but I've been looking into linux and discovered the above. Using @vprodan's wip linux aarch64 sort of works sometimes, but not all detours execute without the KnownMethodThunks implemented. After hackily adding in patterns for the above thunks for net8 by ignoring the bitpacked byte and introducing a lshift of 5 to compensate, things seem to be working there. The shift is actually needed to both adjust for the ignored 3 bits in the bitpacked byte of LDR (literal), and to multiply by 4, since the target address is computed via pc + imm19 * 4. Ignoring those lower 3 bits might actually be generally safe given the thunk alignments, but I'm not entirely sure.

**Comment by owlboy** [comment]
Currently, running the editor in Rosetta is a workaround. Here's some news about the future of Rosetta 2:

https://arstechnica.com/gadgets/2025/06/apple-details-the-end-of-intel-mac-support-and-a-phaseout-for-rosetta-2/
